lib/autokey/scripting/window.py

```diff
@@ -249,3 +302,72 @@ def _run_wmctrl(self, args):
             return 1, 'ERROR: Please install wmctrl'
 
         return returncode, output
+
+    def get_window_list(self, filter_desktop=None):
+        """
+        Returns a list of windows matching an optional desktop filter, requires C{wmctrl}!
+
+        Each list item consists of: C{[hexid, desktop, hostname, title]}
+
+        Where the C{hexid} is the ID used for some other functions (like L{import -window} from ImageMagick).
+
+        C{desktop} is the number of which desktop (sometimes called workspaces) the item appears upon.
+
+        C{hostname} is the hostname of your computer (I'm not sure why this is included as it seems to be the same for everything I have seen)
+
+        C{title} is the title that you would usually see in your window manager of choice.
+
+        @param filter_desktop: String, (usually 0-n) to filter the windows by. Any window not on the given desktop will not be returned.
+        @return: C{[[hexid1, desktop1, hostname1, title1], [hexid2,desktop2,hostname2,title2], ...etc]}
+        Returns C{[]} if no windows are found.
+        """
+        returncode, output = self._run_wmctrl(["-l"])
+        # regex = r"^(0x[0-9a-fA-F]{8})  (\d*) (.*?) (.*?)$"
+        regex = r"^(0x[0-9a-fA-F]{8})\s*(\d*)\s*(.*?)\s(.*?)$"
+        matches = re.findall(regex, output, re.MULTILINE)
+        print(matches)
+        if filter_desktop is None:
+            return matches
+        else:
+            output = []
+            for match in matches:
+                hexid, desktop, hostname, window_title = match
+                if not filter_desktop==desktop:
+                    continue
+                output.append((hexid,desktop,hostname, window_title))
+            return output
+
+    def get_window_hex(self, window_title):
+        """
+        Returns the hexid of the first window to match window_title.
+
+        @param window_title: Window title to match for returning hexid
+        @return: Returns hexid of the window to be used for other functions See L{get_window_geom}, L{visgrep_by_window_id}
+
+        Returns C{None} if no matches are found
+        """
+        windowList = self.get_window_list()
+        for window in windowList:
+            if window_title in window[3]:
+                return window[0]
+        return None
+
+
+
+
+    def get_window_geom(self, window_id):
```

Comment:
How does this function compare to `get_active_geometry` (line 201)? Can you please deduplicate? (Would be nice to have your tests converted for `get_active_geometry` as well)